src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -44,26 +50,47 @@ public override bool Execute()
             {
                 Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
-                using (var httpClient = new HttpClient())
+                if (!DownloadFromUri(Uri, DestinationPath))
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
-                    try
+                    if (!string.IsNullOrWhiteSpace(PrivateUri))
                     {
-                        using (var outStream = File.Create(DestinationPath))
+                        Log.LogMessage(MessageImportance.High, $"Couldn't download file '{Uri}' to '{DestinationPath}'. Trying to download file from '{PrivateUri}' instead.");
+
+                        if (!DownloadFromUri(PrivateUri, DestinationPath))
                         {
-                            getTask.Result.CopyTo(outStream);
+                            Log.LogError($"Couldn't download file '{PrivateUri}' to '{DestinationPath}' either.");
+                            return false;
                         }
                     }
-                    catch (Exception)
+                    else
                     {
-                        File.Delete(DestinationPath);
-                        throw;
+                        Log.LogError($"Couldn't download file '{Uri}' to '{DestinationPath}'");
+                        return false;
                     }
                 }
             }
 
             return true;
         }
+        
+        private bool DownloadFromUri(string source, string target)
+        {
+            try
+            {
+                using (var httpClient = new HttpClient())
+                using (var outStream = File.Create(target))
+                {
+                    var getTask = httpClient.GetStreamAsync(source);
+                    getTask.Result.CopyTo(outStream);
+                }
+
+                return true;
+            }
+            catch (Exception)
```

Comment:
I think it would be better for this code to only catch the exact "OK" case--a 404, right? Rather than going ahead and trying to use internal storage when e.g. the request times out or DNS is bad.

* Retry logic: when you hit a 404, there's no point in retrying the 404, but intermittent network issues should be retried. (I guess this code is missing retry logic overall... but distinguishing 404 makes it easier to implement.)
* In the current world there *shouldn't* be the same files available on both the public and internal storage accounts, but if there for some reason is, having this code only try internal storage after a 404 makes the behavior more predictable. (Super edge case, but something.)